ioGeneratorのテストを乱数を変えながら複数回回すようにする

test/unit/stages.lsでテストするとき、io-generatorがinputに重複があるテストケースを返すと、
io-specのuniqueness of inputを調べているところでコケる。

``` diff
diff --git a/test/unit/stages.ls b/test/unit/stages.ls
index 4da940a..1dd52bf 100644
--- a/test/unit/stages.ls
+++ b/test/unit/stages.ls
@@ -82,8 +82,10 @@ describe 'Stage Data' ->

   describe 'conditional01 stage' ->
     It 'generates 2x if even, 3x otherwise' ->
-      io = conditional01.io-generator @random
+      for i from 1 to 98
+        io = conditional01.io-generator @random

+      io = conditional01.io-generator @random
       expect io .to.satisfy io-spec

       expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
```
とかすると、99回目の乱数を元にio-generatorがされて、
```
 1 failing

  1) Stage Data conditional01 stage generates 2x if even, 3x otherwise:

      AssertionError: expected [ 5, -10, 5, 116 ] to deeply equal [ 5, -10, 116 ]
      + expected - actual

       [
         5
         -10
      -  5
         116
       ]
```
となる。

--- a/stages/conditional01.js
+++ b/stages/conditional01.js
@@ -30,7 +30,7 @@ module.exports = {
 		const input1 = candidates[Math.floor(random() * candidates.length)];
 		const output1 = input1 % 2 === 0 ? 2 * input1 : 3 * input1;
 
-		candidates = candidates.filter((val) => val != input1);
+		candidates = candidates.filter((val) => val !== input1);
 		const input2 = candidates[Math.floor(random() * candidates.length)];
 		const output2 = input2 % 2 === 0 ? 2 * input2 : 3 * input2;
 

--- a/stages/conditional01.js
+++ b/stages/conditional01.js
@@ -24,10 +24,14 @@ module.exports = {
 	input: [5, -10, null, null],
 	output: [15, -20, null, null],
 	ioGenerator: (random) => {
-		const input1 = Math.floor(random() * 200);
+		let candidates = Array.from({length: 200}, (item, index) => index)
+			.filter((val) => val !== 5 && val !== -10);
+
+		const input1 = candidates[Math.floor(random() * candidates.length)];
 		const output1 = input1 % 2 === 0 ? 2 * input1 : 3 * input1;
 
-		const input2 = Math.floor(random() * 200);
+		candidates = candidates.filter((val) => val != input1);
+		const input2 = candidates[Math.floor(random() * candidates.length)];
 		const output2 = input2 % 2 === 0 ? 2 * input2 : 3 * input2;
 
 		return {

--- a/stages/conditional01.js
+++ b/stages/conditional01.js
@@ -1,6 +1,6 @@
 module.exports = {
 	name: 'conditional01',
-	version: 4,
+	version: 5,
 	parts: {
 		wireI: null,
 		wireL: null,

--- a/stages/conditional02.js
+++ b/stages/conditional02.js
@@ -29,7 +29,7 @@ module.exports = {
 
 		const input1 = candidates[Math.floor(random() * candidates.length)];
 
-		candidates = candidates.filter((val) => val != input1);
+		candidates = candidates.filter((val) => val !== input1);
 		const input2 = candidates[Math.floor(random() * candidates.length)];
 
 		return {

--- a/stages/conditional02.js
+++ b/stages/conditional02.js
@@ -1,6 +1,6 @@
 module.exports = {
 	name: 'conditional02',
-	version: 4,
+	version: 5,
 	parts: {
 		wireI: null,
 		wireL: null,

--- a/stages/conditional02.js
+++ b/stages/conditional02.js
@@ -24,10 +24,12 @@ module.exports = {
 	input: [5, 6, null, null],
 	output: [10, 18, null, null],
 	ioGenerator: (random) => {
-		const candidates = Array.from({length: 200}, (item, index) => index + 1)
+		let candidates = Array.from({length: 200 - 1}, (item, index) => index + 1)
 			.filter((val) => val !== 5 && val !== 6);
 
 		const input1 = candidates[Math.floor(random() * candidates.length)];
+
+		candidates = candidates.filter((val) => val != input1);
 		const input2 = candidates[Math.floor(random() * candidates.length)];
 
 		return {

--- a/test/unit/stages.ls
+++ b/test/unit/stages.ls
@@ -61,6 +61,12 @@ io-spec-seq = ({input, output}) ->
     expect i .to.be.an \array .that.has.length-of 2
     expect i.1 .to.be.an \array .that.has.length-of i.0
 
+test-many-times-with-random = (test) ->
+  # 100 is many!
+  for i from 0 to 99
+    random = seedrandom i
+    test random
+
 sum-of-digits = (n) -> n.to-string!split '' .map (parse-int _, 10) .reduce (+)
 
 fibonacci-calc = (n, value = 1, prev = 0) ->
@@ -90,270 +96,287 @@ ten-thousandth-digit-calc = (n) ->
   return x.to-string!split '.' .1.9999 |> parse-int
 
 describe 'Stage Data' ->
-  before-each ->
-    @random = seedrandom ''
-
   describe 'calc03 stage' ->
     It 'generates multiply of 6' ->
-      io = calc03.io-generator @random
+      test-many-times-with-random (random) ->
+        io = calc03.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        6 * input is output
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          6 * input is output
 
-      expect io.input.0 .to.equal 1
-      expect io.input.1 .to.equal 3
+        expect io.input.0 .to.equal 1
+        expect io.input.1 .to.equal 3
 
   describe 'conditional01 stage' ->
     It 'generates 2x if even, 3x otherwise' ->
-      io = conditional01.io-generator @random
+      test-many-times-with-random (random) ->
+        io = conditional01.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is conditional1-calc input
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is conditional1-calc input
 
-      expect io.input.0 .to.equal 5
-      expect io.input.1 .to.equal -10
-      expect io.input.2 .to.be.below 200
-      expect io.input.3 .to.be.below 200
+        expect io.input.0 .to.equal 5
+        expect io.input.1 .to.equal -10
+        expect io.input.2 .to.be.below 200
+        expect io.input.3 .to.be.below 200
 
   describe 'conditional02 stage' ->
     It 'generates 2*5 if 5, 3*6 else if 6, 1 otherwise' ->
-      io = conditional02.io-generator @random
+      test-many-times-with-random (random) ->
+        io = conditional02.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is conditional2-calc input
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is conditional2-calc input
 
-      expect io.input.0 .to.equal 5
-      expect io.input.1 .to.equal 6
-      expect io.input.2 .to.be.below 200
-      expect io.input.3 .to.be.below 200
+        expect io.input.0 .to.equal 5
+        expect io.input.1 .to.equal 6
+        expect io.input.2 .to.be.below 200
+        expect io.input.3 .to.be.below 200
 
   describe 'conditional03 stage' ->
     It 'generates two addition in modulo 7' ->
-      io = conditional03.io-generator @random
+      test-many-times-with-random (random) ->
+        io = conditional03.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is (input + 2) % 7
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is (input + 2) % 7
 
-      expect io.input.0 .to.equal 5
-      expect io.input.1 .to.equal 3
-      expect io.input.2 .to.be.below 200
-      expect io.input.3 .to.be.below 200
-      expect io.input.4 .to.be.below 200
+        expect io.input.0 .to.equal 5
+        expect io.input.1 .to.equal 3
+        expect io.input.2 .to.be.below 200
+        expect io.input.3 .to.be.below 200
+        expect io.input.4 .to.be.below 200
 
   describe 'factorial stage' ->
     It 'generates factorials' ->
-      io = factorial.io-generator @random
+      test-many-times-with-random (random) ->
+        io = factorial.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is mathjs.factorial input
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is mathjs.factorial input
 
-      expect io.input.3 .to.equal 15
+        expect io.input.3 .to.equal 15
 
   describe 'parity stage' ->
     It 'generates parities' ->
-      io = parity.io-generator @random
+      test-many-times-with-random (random) ->
+        io = parity.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        sum-of-digits(input) % 10 is output
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          sum-of-digits(input) % 10 is output
 
-      expect sum-of-digits io.input.0 .to.be.below 10
-      expect sum-of-digits io.input.1 .to.be.least 10 .and.below 20
-      expect sum-of-digits io.input.2 .to.be.least 20
+        expect sum-of-digits io.input.0 .to.be.below 10
+        expect sum-of-digits io.input.1 .to.be.least 10 .and.below 20
+        expect sum-of-digits io.input.2 .to.be.least 20
 
   describe 'fibonacci stage' ->
     It 'generates fibonaccis' ->
-      io = fibonacci.io-generator @random
+      test-many-times-with-random (random) ->
+        io = fibonacci.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is fibonacci-calc input
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is fibonacci-calc input
 
-      expect io.input.3 .to.equal 20
+        expect io.input.3 .to.equal 20
 
   describe 'power-hard stage' ->
     It 'generates powers' ->
-      io = power-hard.io-generator @random
+      test-many-times-with-random (random) ->
+        io = power-hard.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is Math.pow input.0, input.1
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is Math.pow input.0, input.1
 
-      # TODO: test input.0 ond input.1
-      expect io.input.2 .to.deep.equal [2, 35]
+        # TODO: test input.0 ond input.1
+        expect io.input.2 .to.deep.equal [2, 35]
 
   describe 'gcd stage' ->
     It 'generates GCDs' ->
-      io = gcd.io-generator @random
+      test-many-times-with-random (random) ->
+        io = gcd.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is mathjs.gcd input.0, input.1
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is mathjs.gcd input.0, input.1
 
-      expect io.input.3 .to.deep.equal [56, 1275]
+        expect io.input.3 .to.deep.equal [56, 1275]
 
   describe 'lcm stage' ->
     It 'generates LCMs' ->
-      io = lcm.io-generator @random
+      test-many-times-with-random (random) ->
+        io = lcm.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is mathjs.lcm input.0, input.1
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is mathjs.lcm input.0, input.1
 
-      expect io.input.3 .to.deep.equal [450, 756]
+        expect io.input.3 .to.deep.equal [450, 756]
 
   describe 'sqrt-easy stage' ->
     It 'generates square roots' ->
-      io = sqrt-easy.io-generator @random
+      test-many-times-with-random (random) ->
+        io = sqrt-easy.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        input is output ** 2
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          input is output ** 2
 
-      expect io.output.0 .to.be.least 2 .and.most 14
-      expect io.output.1 .to.be.least 2 .and.most 14
-      expect io.output.2 .to.be.least 2 .and.most 14
-      expect io.output.3 .to.equal 15
+        expect io.output.0 .to.be.least 2 .and.most 14
+        expect io.output.1 .to.be.least 2 .and.most 14
+        expect io.output.2 .to.be.least 2 .and.most 14
+        expect io.output.3 .to.equal 15
 
   describe 'sqrt-hard stage' ->
     It 'generates hard square roots' ->
-      io = sqrt-hard.io-generator @random
+      test-many-times-with-random (random) ->
+        io = sqrt-hard.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        input is output ** 2
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          input is output ** 2
 
-      expect io.output.0 .to.equal 3
-      expect io.output.1 .to.be.least 10 .and.below 100
-      expect io.output.2 .to.be.least 1000 .and.below 5000
-      expect io.output.3 .to.equal 5103
+        expect io.output.0 .to.equal 3
+        expect io.output.1 .to.be.least 10 .and.below 100
+        expect io.output.2 .to.be.least 1000 .and.below 5000
+        expect io.output.3 .to.equal 5103
 
   describe 'msd stage' ->
     It 'generates MSDs' ->
-      io = msd.io-generator @random
+      test-many-times-with-random (random) ->
+        io = msd.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is parse-int input.to-string!0
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is parse-int input.to-string!0
 
-      expect io.input.2 .to.equal 0
-      expect io.input.3 .to.equal 123456789
+        expect io.input.2 .to.equal 0
+        expect io.input.3 .to.equal 123456789
 
   describe 'mod3-hard stage' ->
     It 'generates moduloes of 3' ->
-      io = mod3-hard.io-generator @random
+      test-many-times-with-random (random) ->
+        io = mod3-hard.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is input % 3
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is input % 3
 
-      expect io.input.0 .to.be.least 10000000 .and.below 90000000
-      expect io.input.1 .to.be.least 10000000 .and.below 90000000
-      expect io.input.2 .to.be.least 10000000 .and.below 90000000
-      expect io.input.3 .to.be.least 10000000 .and.below 90000000
-      expect io.input.4 .to.be.least 99990000 .and.below 100000000
+        expect io.input.0 .to.be.least 10000000 .and.below 90000000
+        expect io.input.1 .to.be.least 10000000 .and.below 90000000
+        expect io.input.2 .to.be.least 10000000 .and.below 90000000
+        expect io.input.3 .to.be.least 10000000 .and.below 90000000
+        expect io.input.4 .to.be.least 99990000 .and.below 100000000
 
-      expect io.output.0 .to.equal 0
-      expect io.output.1 .to.equal 1
-      expect io.output.2 .to.equal 2
+        expect io.output.0 .to.equal 0
+        expect io.output.1 .to.equal 1
+        expect io.output.2 .to.equal 2
 
   describe 'the-fifth-max stage' ->
     It 'generates medians' ->
-      io = the-fifth-max.io-generator @random
+      test-many-times-with-random (random) ->
+        io = the-fifth-max.io-generator random
 
-      expect io .to.satisfy io-spec-seq
+        expect io .to.satisfy io-spec-seq
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is mathjs.median input.1 and input.1.filter (is output) .length is 1
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is mathjs.median input.1 and input.1.filter (is output) .length is 1
 
-      expect io.input.0 .to.deep.equal [5, [1, 2, 3, 4, 5]]
+        expect io.input.0 .to.deep.equal [5, [1, 2, 3, 4, 5]]
 
-      expect io.input.1.0 .to.equal 5
-      expect io.input.2.0 .to.equal 7
-      expect io.input.3.0 .to.equal 7
-      expect io.input.4.0 .to.equal 7
+        expect io.input.1.0 .to.equal 5
+        expect io.input.2.0 .to.equal 7
+        expect io.input.3.0 .to.equal 7
+        expect io.input.4.0 .to.equal 7
 
-      expect io.input.4.1.6 .to.equal mathjs.median io.input.4.1
+        expect io.input.4.1.6 .to.equal mathjs.median io.input.4.1
 
   describe 'repeat-self stage' ->
     It 'generates valid io' ->
-      io = repeat-self.io-generator @random
+      test-many-times-with-random (random) ->
+        io = repeat-self.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is (input.to-string!repeat input |> parse-int)
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is (input.to-string!repeat input |> parse-int)
 
-      expect io.input.0 .to.equal 1
-      expect io.input.3 .to.equal 9
+        expect io.input.0 .to.equal 1
+        expect io.input.3 .to.equal 9
 
   describe 'fibonacci-hord stage' ->
     It 'generates fibonaccis' ->
-      io = fibonacci-hard.io-generator @random
+      test-many-times-with-random (random) ->
+        io = fibonacci-hard.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is fibonacci-calc input
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is fibonacci-calc input
 
-      expect io.input.3 .to.equal 30
+        expect io.input.3 .to.equal 30
 
   describe 'factorization stage' ->
     It 'generates factorization' ->
-      io = factorization.io-generator @random
+      test-many-times-with-random (random) ->
+        io = factorization.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is factorization-calc input
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is factorization-calc input
 
-      expect io.input.3 .to.equal 211
-      expect io.input.4 .to.equal 221
+        expect io.input.3 .to.equal 211
+        expect io.input.4 .to.equal 221
 
   describe 'spaceship stage' ->
     It 'generates spaceship' ->
-      io = spaceship.io-generator @random
+      test-many-times-with-random (random) ->
+        io = spaceship.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        output is spaceship-calc ...input
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          output is spaceship-calc ...input
 
-      expect io.output.0 .to.equal 0
-      expect io.output.1 .to.equal -1
-      expect io.output.2 .to.equal 1
+        expect io.output.0 .to.equal 0
+        expect io.output.1 .to.equal -1
+        expect io.output.2 .to.equal 1
 
   describe '10000th-digit stage' ->
     It 'generates 10000th-digit' ->
-      io = ten-thousandth-digit.io-generator @random
+      test-many-times-with-random (random) ->
+        io = ten-thousandth-digit.io-generator random
 
-      expect io .to.satisfy io-spec
+        expect io .to.satisfy io-spec
 
-      # Uniqueness of output
-      unique-output = Array.from new Set io.output
-      expect io.output .to.deep.equal unique-output
+        # Uniqueness of output
+        unique-output = Array.from new Set io.output
+        expect io.output .to.deep.equal unique-output
 
-      expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
-        mathjs.is-prime input
-        and output is ten-thousandth-digit-calc input
+        expect zip io.input, io.output .to.all.satisfy ([input, output]) ->
+          mathjs.is-prime input
+          and output is ten-thousandth-digit-calc input
 
-      expect io.input.4 .to.equal 107
+        expect io.input.4 .to.equal 107

--- a/test/unit/stages.ls
+++ b/test/unit/stages.ls
@@ -63,7 +63,7 @@ io-spec-seq = ({input, output}) ->
 
 test-many-times-with-random = (test) ->
   # 100 is many!
-  for i from 0 til 100 - 1
+  for i from 0 til 100
     random = seedrandom i
     test random
 

--- a/test/unit/stages.ls
+++ b/test/unit/stages.ls
@@ -63,7 +63,7 @@ io-spec-seq = ({input, output}) ->
 
 test-many-times-with-random = (test) ->
   # 100 is many!
-  for i from 0 to 99
+  for i from 0 to 100 - 1
     random = seedrandom i
     test random
 

--- a/test/unit/stages.ls
+++ b/test/unit/stages.ls
@@ -63,7 +63,7 @@ io-spec-seq = ({input, output}) ->
 
 test-many-times-with-random = (test) ->
   # 100 is many!
-  for i from 0 to 100 - 1
+  for i from 0 til 100 - 1
     random = seedrandom i
     test random
 

